Teach ZeroOrderHold to handle AbstractValue's.

EricCousineau-TRI opened the pull request:
This resolves the main intent of #6460 by permitting `ZeroOrderHold` to handle abstract values.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6488)
<!-- Reviewable:end -->

siyuanfeng-tri commented:
+@siyuanfeng-tri for feature review. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-:-KntX2ifb8aaSstkKpUw:b3988e4)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuM5BqrXbSELOt99iq:-KnuM5Br2uTkOBooRBoe:bu8ifh7) ([raw file](https://github.com/robotlocomotion/drake/blob/e307df9bd66f07ffbd4e589135d7e088047dc3be/drake/systems/primitives/zero_order_hold-inl.h#L38)):*
> ```objc
>   this->DeclareAbstractInputPort(model_value);
>   // We must bind these because the instance method pointer overload expects
>   // the allocator to return a value-type `OutputType`, which would cause
> ```

BTW i thought the issue is you don't know what underlying type model_value has since it's type erased. 

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuMltgO5Ql6aEQNhBf:-KnuMltgO5Ql6aEQNhBg:b-db3qlx) ([raw file](https://github.com/robotlocomotion/drake/blob/e307df9bd66f07ffbd4e589135d7e088047dc3be/drake/systems/primitives/zero_order_hold-inl.h#L79)):*
> ```objc
>   DRAKE_ASSERT(is_abstract());
>   // Do not use template get_abstract_state<U>, because this will attempt to
>   // cast the value to Value<AbstractValue>. Instead, query the AbstractValue
> ```

BTW I found this comment confusing. I think @jwnimmer-tri explicitly wanted to prevent Value<AbsValue> at some point. 
In any event, my interpretation is that you don't have a type, so you are constrained to use AbstrasctValue::SetFrom 

---

*[drake/systems/primitives/zero_order_hold.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuI3yXr8spBaa9LCFm:-KnuI3yXr8spBaa9LCFn:bfkzoj6) ([raw file](https://github.com/robotlocomotion/drake/blob/e307df9bd66f07ffbd4e589135d7e088047dc3be/drake/systems/primitives/zero_order_hold.h#L38)):*
> ```objc
>   // Reflect SisoVectorSystem API.
>   // TODO(eric.cousineau): Create a SisoSystem that is type-agnostic, and
>   // have both this and SisoVectorSystem inherit from this.
> ```

from it.

---

*[drake/systems/primitives/test/zero_order_hold_test.cc, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuJYe72bfEzz6Vdkxs:-KnuJYe72bfEzz6Vdkxt:b-k7ddhl) ([raw file](https://github.com/robotlocomotion/drake/blob/e307df9bd66f07ffbd4e589135d7e088047dc3be/drake/systems/primitives/test/zero_order_hold_test.cc#L112)):*
> ```cpp
> 
> // Tests that the zero-order hold has discrete state.
> TEST_P(ZeroOrderHoldTest, ReservesState) {
> ```

BTW, i like InitialState better. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-:-KnuNk2KjfWb4N5eVg1h:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuM5BqrXbSELOt99iq:-KnuPl0xukELX_cseC98:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e307df9bd66f07ffbd4e589135d7e088047dc3be/drake/systems/primitives/zero_order_hold-inl.h#L38)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW i thought the issue is you don't know what underlying type model_value has since it's type erased. 
</blockquote></details>

Done.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 79 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuMltgO5Ql6aEQNhBf:-KnuPcrXNScRbIqJL-O4:b-u670eg) ([raw file](https://github.com/robotlocomotion/drake/blob/e307df9bd66f07ffbd4e589135d7e088047dc3be/drake/systems/primitives/zero_order_hold-inl.h#L79)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW I found this comment confusing. I think @jwnimmer-tri explicitly wanted to prevent Value<AbsValue> at some point. 
In any event, my interpretation is that you don't have a type, so you are constrained to use AbstrasctValue::SetFrom 
</blockquote></details>

Done. My intent was to capture just that in the comment (that `Value<AbstractValue>` is an invalid type by design) - tried to clarify the comment a little more.

---

*[drake/systems/primitives/zero_order_hold.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuI3yXr8spBaa9LCFm:-KnuPiWLjZs6N4X98CaT:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/e307df9bd66f07ffbd4e589135d7e088047dc3be/drake/systems/primitives/zero_order_hold.h#L38)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

from it.
</blockquote></details>

Done. Thanks!

---

*[drake/systems/primitives/test/zero_order_hold_test.cc, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuJYe72bfEzz6Vdkxs:-KnuPjZNtYlp3f5lfia3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e307df9bd66f07ffbd4e589135d7e088047dc3be/drake/systems/primitives/test/zero_order_hold_test.cc#L112)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

BTW, i like InitialState better. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
+@jwnimmer-tri for platform review, please.
(Please let me know if you would like me to re-assign to next Wednesday's platform reviewer given the holidays.)

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-:-KnuPoHxYKfm7VU6lskD:b-7dhz5t)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 3 of 4 files at r2.
Review status: 3 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 9 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuTDgRZR1ECokfoVzw:-KnuTDgSPR5hDo5p1loO:bl9o1hl) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L9)):*
> ```objc
> 
> #include <memory>
> #include <vector>
> ```

FYI These are a bit gratuitous, but not defective.  This is the "related" implementation file for `zero_order_hold.h` so can presume all of those includes.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 52 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuVCNbfoqIBiNwatR-:-KnuVCNcPsG8qk12ql_a:b-k687ep) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L52)):*
> ```objc
>       BasicVector<T>* output) const {
>   DRAKE_ASSERT(!is_abstract());
>   const auto& state_value = *context.get_discrete_state(0);
> ```

`auto` is too obscure; the type is not obvious from RHS code.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 79 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuVT2HRLPPI9HCrecc:-KnuVT2I6ZTHT4-6Ghfh:b-k687ep) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L79)):*
> ```objc
>   // Do not use `get_abstract_state<AbstractValue>` since it would cast
>   // the value to `Value<AbstractValue>`, which is an invalid type by design.
>   const auto& state_value =
> ```

`auto` is too obscure; the type is not obvious from RHS code.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 90 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuVUBn5BiCEUg87HZv:-KnuVUBoTXC0KqIS5heC:b-h6rwnv) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L90)):*
> ```objc
>     State<T> *state) const {
>   DRAKE_ASSERT(is_abstract());
>   const auto& input_value = *this->EvalAbstractInput(context, 0);
> ```

`auto` is too obscure; the type is not obvious from RHS code.
Ditto for the next statement.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuUKq-epsTdyaZh0pa:-KnuUKq-epsTdyaZh0pb:bfs9xtb) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L106)):*
> ```objc
>     // Transmographication not supported for abstract values.
>     throw std::runtime_error(
>         "DoToSymbolic not implemented for abstract values.");
> ```

This needs to be mentioned in the class Doxygen somewhere so users are aware.  (Or else possibly `UninterpretedFunction` could be used to make it work.)

---

*[drake/systems/primitives/zero_order_hold.h, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuRp2nMfzKNy29-yUX:-KnuRp2oac6jM7bnv_2-:b43794p) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold.h#L36)):*
> ```objc
>   ~ZeroOrderHold() override {}
> 
>   // Reflect SisoVectorSystem API.
> ```

What does "Reflect" mean in this situation?  (Or perhaps: I'm not sure this line is adding any value, since I can't come up with a possible meaning here that's useful.)

The `get_input_port()` and `get_output_port()` are a widely used convention, beyond just SisoVectorSystem's particular implementation of the pattern.

---

*[drake/systems/primitives/zero_order_hold.h, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuViIEdCgoNhIom8eW:-KnuViIFRluetbuKWAFg:bids6si) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold.h#L92)):*
> ```objc
> 
>   const double period_sec_{};
>   const std::unique_ptr<AbstractValue> abstract_model_value_;
> ```

FYI I think `const std::unique_ptr<const AbstractValue>` works, if you prefer.

---

*[drake/systems/primitives/test/zero_order_hold_test.cc, line 35 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuWgILO-hcI7b_W2WP:-KnuWgILO-hcI7b_W2WQ:b-lzua78) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/test/zero_order_hold_test.cc#L35)):*
> ```cpp
> void CheckForSinglePeriodicEvent(const EventListType& events) {
>   EXPECT_EQ(events.size(), 1);
>   EXPECT_EQ(events.front()->get_trigger_type(),
> ```

FYI Ideally, the test code would be written so that if `events.empty()` the `.front()` call here wouldn't attempted (and probably segfault).

---

*[drake/systems/primitives/test/zero_order_hold_test.cc, line 103 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuX8ccqyB3taBnX4vm:-KnuX8cdhoCsAqSevfYl:b-3v194s) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/test/zero_order_hold_test.cc#L103)):*
> ```cpp
>     // `DoToSymbolic` is not supported in this case.
>     EXPECT_FALSE(hold_->HasAnyDirectFeedthrough());
>   }
> ```

It would be better to still have `HasAnyDirectFeedthrough` pass; either by making symbolic work, or by using the opt-out APIs in System (or LeafSystem?  I forget) to directly state our feedthrough.

---

*[drake/systems/primitives/test/zero_order_hold_test.cc, line 107 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuXU7ZiVrf5PI6f2eF:-KnuXU7ZiVrf5PI6f2eG:bm8lj4r) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/test/zero_order_hold_test.cc#L107)):*
> ```cpp
> 
> template <typename Derived>
> bool all_nan(const Eigen::MatrixBase<Derived>& X) {
> ```

Dead code?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 1 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 9 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuTDgRZR1ECokfoVzw:-KnusqXZXT90Ca1VWWxr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI These are a bit gratuitous, but not defective.  This is the "related" implementation file for `zero_order_hold.h` so can presume all of those includes.
</blockquote></details>

Done.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 52 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuVCNbfoqIBiNwatR-:-Knusq5Rqw94tERSZGCP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L52)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`auto` is too obscure; the type is not obvious from RHS code.
</blockquote></details>

Done.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 79 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuVT2HRLPPI9HCrecc:-KnuspZtqSKvVGXXu1_-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L79)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`auto` is too obscure; the type is not obvious from RHS code.
</blockquote></details>

Done.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 90 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuVUBn5BiCEUg87HZv:-KnusoO8g_Q18TZPxHzw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L90)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`auto` is too obscure; the type is not obvious from RHS code.
Ditto for the next statement.
</blockquote></details>

Done.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 106 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuUKq-epsTdyaZh0pa:-KnusZE6D1tj2LrJWClv:bb32yx1) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold-inl.h#L106)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This needs to be mentioned in the class Doxygen somewhere so users are aware.  (Or else possibly `UninterpretedFunction` could be used to make it work.)
</blockquote></details>

Done. Added a `@note` to Doxygen.

---

*[drake/systems/primitives/zero_order_hold.h, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuRp2nMfzKNy29-yUX:-Knunyfi6-uxVDnu7iNU:b8u6xp7) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold.h#L36)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

What does "Reflect" mean in this situation?  (Or perhaps: I'm not sure this line is adding any value, since I can't come up with a possible meaning here that's useful.)

The `get_input_port()` and `get_output_port()` are a widely used convention, beyond just SisoVectorSystem's particular implementation of the pattern.
</blockquote></details>

Removed the comment.

Looking at other usages of `get_input_port()` and `get_output_port()`, it seems like some of them should possibly be inheriting from `Af` (such as `AffineSystem`, `Saturation`, etc).
(That being said, I am not sure if I understand the rationale on hiding zero-width ports in `AffineSystem`, but that may be another discussion.)

---

*[drake/systems/primitives/zero_order_hold.h, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuViIEdCgoNhIom8eW:-Knup3zPd1RVLZMyqk52:b-8us15e) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/zero_order_hold.h#L92)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I think `const std::unique_ptr<const AbstractValue>` works, if you prefer.
</blockquote></details>

Done. Definitely prefer that one!

---

*[drake/systems/primitives/test/zero_order_hold_test.cc, line 35 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuWgILO-hcI7b_W2WP:-Knut379h9KIFNBvHbux:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/test/zero_order_hold_test.cc#L35)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Ideally, the test code would be written so that if `events.empty()` the `.front()` call here wouldn't attempted (and probably segfault).
</blockquote></details>

Done.

---

*[drake/systems/primitives/test/zero_order_hold_test.cc, line 103 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuX8ccqyB3taBnX4vm:-Knut7s61u0cW41DJU-D:b-kwviq5) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/test/zero_order_hold_test.cc#L103)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It would be better to still have `HasAnyDirectFeedthrough` pass; either by making symbolic work, or by using the opt-out APIs in System (or LeafSystem?  I forget) to directly state our feedthrough.
</blockquote></details>

Done. Thanks for pointing that out!

---

*[drake/systems/primitives/test/zero_order_hold_test.cc, line 107 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-KnuXU7ZiVrf5PI6f2eF:-Knut9NtqiiQ7tOVYNJV:b-urihqt) ([raw file](https://github.com/robotlocomotion/drake/blob/1fa2ba1adc22c6c4c45c663806c823b59924776e/drake/systems/primitives/test/zero_order_hold_test.cc#L107)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Dead code?
</blockquote></details>

Done. Sorry about that!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-Knuvy74I9cOXtODqFKS:-Knuvy75vN4C5Qa6jxJm:b-cu3h79) ([raw file](https://github.com/robotlocomotion/drake/blob/4bb3231e8820bcda193f12174f67650a9e9e3847/drake/systems/primitives/zero_order_hold-inl.h#L112)):*
> ```objc
>   } else {
>     // Return `nullptr` to enable control to reach `DoHasDirectFeedthrough`.
>     // See Doxygen comments regarding transmografication.
> ```

BTW Typo `gri` not `gra`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-:-KnuxFypxYWkR-VoPZLn:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Aha.  Those includes in the inl file that I nitpicked were there because of cpplint's broken IWYU implementation; cpplint is angry with this PR now.

siyuanfeng-tri commented:
Reviewed 1 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Oops! Fixed that.

---

Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/primitives/zero_order_hold-inl.h, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-Knuvy74I9cOXtODqFKS:-KohX3X558zWupMasM6j:boyexb3) ([raw file](https://github.com/robotlocomotion/drake/blob/4bb3231e8820bcda193f12174f67650a9e9e3847/drake/systems/primitives/zero_order_hold-inl.h#L112)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Typo `gri` not `gra`.
</blockquote></details>

Done. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488#-:-KohX0O_n-6-t6Dls0q_:bk3xyms)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6488)*
<!-- Sent from Reviewable.io -->
